File: built-in plugins dsp/mb-wt-supersaw.cpp
/**
 * MB WT Supersaw
 * Category : instrument
 * Type     : wavetable
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Wavetable supersaw stack
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_WT_SUPERSAW_H
#define MB_WT_SUPERSAW_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbWtSupersaw : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-wt-supersaw";
    static constexpr const char* PLUGIN_NAME    = "MB WT Supersaw";
    static constexpr const char* PLUGIN_TYPE    = "wavetable";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float position = 0.4f;  // range [0, 1]
    float detune = 10f;  // range [0, 50]
    float volume = 0.75f;  // range [0, 1]
    };

    MbWtSupersaw() = default;
    ~MbWtSupersaw() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.position = std::clamp(params.position, 0f, 1f);
        params.detune = std::clamp(params.detune, 0f, 50f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB WT Supersaw
        return input;
    }
};

#endif // MB_WT_SUPERSAW_H

File: built-in plugins dsp/mb-wt-supersaw.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbWtSupersawPlugin: PluginDefinition = { id: 'mb-wt-supersaw', slug: 'mb-wt-supersaw', name: 'MB WT Supersaw', category: 'instrument', type: 'wavetable', version: '1.0.0', description: 'Wavetable supersaw stack', author: 'Max Booster', oscillators: [{ type: 'sawtooth', detune: -10, gain: 0.35 }, { type: 'sawtooth', detune: 10, gain: 0.35 }], envelope: { attack: 0.02, decay: 0.4, sustain: 0.7, release: 0.5 }, parameters: [{ id: 'position', name: 'Wave Position', type: 'float', defaultValue: 0.4, minValue: 0, maxValue: 1, automatable: true }, { id: 'detune', name: 'Detune', type: 'float', defaultValue: 10, minValue: 0, maxValue: 50, automatable: true }, { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.75, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { position: 0.4, detune: 10, volume: 0.75 } };

export default MbWtSupersawPlugin;
